nomad/job_endpoint.go

```diff
@@ -1785,7 +1785,7 @@ func (j *Job) Plan(args *structs.JobPlanRequest, reply *structs.JobPlanResponse)
 	}
 
 	// Create the scheduler and run it
-	sched, err := scheduler.NewScheduler(eval.Type, j.logger, snap, planner)
+	sched, err := scheduler.NewScheduler(eval.Type, j.logger, nil, snap, planner)
```

Comment:
If we made this something like `type BadNodeReporter interface { Report(badNodes *T) }` we could have an alternative implementation in `Job.Plan` for reporting bad nodes immediately to users.

Not sure it's worth it, at least for v1. The ability to get immediate feedback vs grepping logs seems incredibly useful. However, it's not at all what the `Job.Plan` endpoint is for. Telling a job submitter about a bad node absolutely breaks the principle of least surprise.

To begin with giving `Job.Plan` the same implementation as the scheduler might be nice. That way users could (hopefully) trigger the log message and metric without actually having to schedule anything.

Reply:
By adding the collision to the scheduler metrics we don't need to listen for individual events here. The result of a failed `Job.Plan` looks like this:

```
$ nomad job plan example.nomad
Job: "example"
Task Group: "cache" (1 create)
  Task: "redis"

Scheduler dry-run:
- WARNING: Failed to place all allocations.
  Task Group "cache" (failed to place 1 allocation):
    * Resources exhausted on 1 nodes
    * Dimension "network: collision when reserving port range for node 97ba7419-d875-e370-11b6-ad6d594f724f: [invalid port 808081]" exhausted on 1 nodes

Job Modify Index: 10
To submit the job with version verification run:

nomad job run -check-index 10 example.nomad

When running the job with the check-index flag, the job will only be run if the
job modify index given matches the server-side version. If the index has
changed, another user has modified the job and the plan's results are
potentially invalid.
```

Reply:
Reading the code again, I changed to metric key here to just `network: port collision`. The reason for that is that using `reason` directly would result in unique key values, so for large clusters this would result in several `Dimension "network: collision...` messages here.

One option could be to special case this metric and output more information (`refer to server logs for more information` or something like that) in the command directly?

Reply:
https://github.com/hashicorp/nomad/pull/11793/commits/eed3254fb9882f2fe4e5605f9bda42ed325ee707 adds some logic to display and (optionally) log scheduler events.